specs/WinRT/WinRTAPIContracts.md

```diff
@@ -0,0 +1,78 @@
+
+
+
+# Windows App SDK API Contracts
+
+## Summary
+Customers trust that Microsoft products are sound.  As Windows App SDK evolves and more developers use Windows App SDK in their own projects Windows App SDK needs a way, at build time and run time, to allow apps to light up API features and to gracefully degrade.  This can be achieved by looking at the deltas between APIs.  
+
+In order to allow apps to depend on API versions Windows App SDK will add contracts to all their WinRT APIs.
+
+
+## Architecture Decisions
+1. Contract boundaries
+   1. One contract per feature/component/sub-system
+   2. Contracts will not span across a transport package/repo.
+2. Contract numbering.
+Contracts will be versioned with one number, starting with 1 and is incremented for any interface change.  This version does not correlate with the Windows App SDK version.
+3. Enforcement
+The build systems will have enforcement gates to make sure that contracts are present, up to date, and there are no conflicts.
+4. Contract Migration.
+Contract Migration will not be a part of this spec.  Windows App SDK will add a spec on contract migration if, and when, there is a situation where it is needed.
+5. Internal contracts
+All WinRT internal APIs will have an contract and the version will always be 1 regardless of any interface changes.
+
+## Adding a new contract
+Developers are encouraged to use their best judgment when deciding where to place a new contract.  If in doubt you can reach out to the Windows App SDK team.
+
+## Example
+
+    namespace Microsoft.Windows.System
+    {
+        [contractversion(2)]
+        apicontract MicrosoftWindowsSystem
+        
+        // version 2
+        [Contract(MicrosoftWindowsSystem(2)]
+        public enum Scope
+        {
+            user,
+            machine
+        }
+
+        [Contract(MicrosoftWindowsSystem(2)]
+        public class EnvironmentManager
+        {
+            public String GetEnvironmentVariable(String name, Scope scope);
+        }
+
+        // base version
+        [contract(MicrosoftWindowsSystem, 1)]
+        public class EnvironmentManager
+        {
+            public String GetEnvironmentVariable(String name);
+        }
+    }
+
+## Forbidden interface changes
+To help keep ABI and binary compatibility between different versions of Windows App SDK here is a list of changes that should not happen between interface versions.
+1. Method order in the interface.
+2. Method signatures.
+3. Structs.
```

Comment:
Structs cannot be modified. Interfaces cannot be modified. Enums can only be added to, but existing values and names cannot change. 

We should also make a decision about runtimeclass derivation. If I have runtimeclass A derives from B and then I introduce runtimeclass C that derives from B, can I change A to derive from C later? In theory this is safe but we have avoided doing it to date.

Reply:
I'll also say that many people don't grok that a runtimeclass is an aggregation of interfaces, so if you want this to be a doc that folks reference when they are trying to figure out how to manage their contract versioning, please include examples like the internal wiki page does. Most additive things are possible but the syntax for doing it isn't always obvious, like the contract block syntax or adding a setter to a property that previously only had a getter.

Reply:
> want this to be a doc that folks reference

Thank you for bringing this issue up.  This is a document for specs, not coding guidelines.  I'll make a separate PR with contract guidelines, including forbidden changes.  For this PR This section will be removed because "Don't do these things" is more "Coding guidelines" and not "spec"